### PR TITLE
dialects: Make arith.BinaryOp Generic.

### PR DESCRIPTION
**Pull request by PapyChacal**
Use the new Generic Operation base class possibility brought by #979, make BinaryOperation generic, delete a third of `arith.py` :tada: 
I'm going to chop the rest of it soon!

### CODE REVIEW

File: xdsl/dialects/arith.py [hunk 1]
```diff
@@ -140,31 +140,41 @@ def from_float_and_width(
...
+    traits = frozenset([Pure()])
+
+    lhs: Annotated[Operand, _T]
+    rhs: Annotated[Operand, _T]
+    result: Annotated[OpResult, _T]
+
+    @classmethod
+    def get(
+        cls,
+        operand1: Union[Operation, SSAValue], operand2: Union[Operation, SSAValue]
```
Comment (AntonLydike) [inline xdsl/dialects/arith.py]: Instead of `Union[A, B]` i would suggest `A | B`. It's more in line with the rest of the repo.

**Comment by PapyChacal** [comment]
:warning: I factorized Pure to BinaryOperation, But didn't check it should be for every one of them yet, input appreciated if somebody already knows!

**Review by AntonLydike — APPROVED** [review]
:rocket:

**Comment by PapyChacal** [comment]
This reduces coverage though :frowning_face:

**Comment by georgebisbas** [comment]
> This reduces coverage though frowning_face

Do not get fooled, you did the best practice here!

**Review by georgebisbas — APPROVED** [review]
Top

File: xdsl/dialects/arith.py [hunk 2]
```diff
@@ -141,29 +141,42 @@ def from_float_and_width(
         return Constant.create(result_types=[typ], attributes={"value": val})
 
 
-class BinaryOperation(IRDLOperation):
+_T = TypeVar("_T")
```
Comment (math-fehr) [inline xdsl/dialects/arith.py:144]:
```suggestion
_T = TypeVar("_T", bound=Attribute)
```
So people writing `BinaryOperation[int]` will get an error.

File: xdsl/dialects/arith.py [hunk 3]
```diff
@@ -141,29 +141,42 @@ def from_float_and_width(
...
+
+
+class BinaryOperation(IRDLOperation, Generic[_T]):
     """A generic operation. Operation definitions inherit this class."""
```
Comment (math-fehr) [inline xdsl/dialects/arith.py:148]: Could we maybe rename this?

**Review by math-fehr — APPROVED** [review]
Very nice!
For the `Pure` aspect, it depends if we want to match MLIR notion of `Pure` (which we will want to match at some point).
All these operations are pure in the sense that their result only depend from their operands. However, they might trigger UB, so it is not safe to hoist them outside of a loop, so they are not pure in that sense.
MLIR distinguish both by having the `NoSideEffects` trait and the `CanSafelySpeculate` trait (Pure being the combination of both). Here, all operations are `NoSideEffects`, but they would not be `CanSafelySpeculate`.
If we decide to not model that yet, then let's mark them all pure. However, we need to keep in mind that we will have to change at some point.

**Comment by PapyChacal** [comment]
Interesting! Then I'll just move the trait back to arith.addi only as it was before the PR, to not generalize this as it was probably just a first experiment with the new xDSL traits
I'm not after this for now and it seems like deserving a proper discussion and PR!
Thanks for your input 🙂

File: xdsl/dialects/arith.py (hunk 3 above)
Comment (PapyChacal) [reply xdsl/dialects/arith.py]: @math-fehr I tried something, but I'm not sure how much we want to phrase here. WDYT?

**Pull request merged**